File: Web_Stadium/Services/StandingService.cs
using Microsoft.EntityFrameworkCore;
using Web_Stadium.EFCore;

namespace Web_Stadium.Services
{
    /// <summary>
    /// Tính bảng xếp hạng theo chuẩn FIFA tie-breakers
    /// View chỉ render, không tính toán
    /// </summary>
    public class StandingService
    {
        private readonly SanBongContext _context;
        public StandingService(SanBongContext context) => _context = context;

        // ══════════════════════════════════════════════════════════
        // Lấy BXH toàn giải — trả về dict {bangId: [rows]}
        // ══════════════════════════════════════════════════════════
        public async Task<Dictionary<int, List<StandingRow>>> GetStandings(int giaiDauId)
        {
            var giai = await _context.GiaiDaus
                .Include(g => g.BangDaus)
                .Include(g => g.DoiBongs)
                .Include(g => g.TranDaus)
                .FirstOrDefaultAsync(g => g.Id == giaiDauId);

            if (giai == null) return new();

            var result = new Dictionary<int, List<StandingRow>>();

            foreach (var bang in giai.BangDaus)
            {
                var doiTrongBang = giai.DoiBongs
                    .Where(d => d.BangId == bang.Id).ToList();

                var stats = doiTrongBang.ToDictionary(d => d.Id, d => new StandingRow
                {
                    DoiId = d.Id,
                    TenDoi = d.TenDoi,
                    LogoUrl = d.LogoUrl
                });

                // Tính từ các trận đã Closed
                var transClosed = giai.TranDaus
                    .Where(t => t.BangId == bang.Id && t.TrangThai == "Closed")
                    .ToList();

                foreach (var tran in transClosed)
                {
                    if (!tran.BanThangNha.HasValue || !tran.BanThangKhach.HasValue) continue;
                    if (!stats.ContainsKey(tran.DoiNhaId) ||
                        !stats.ContainsKey(tran.DoiKhachId)) continue;

                    var nha = stats[tran.DoiNhaId];
                    var khach = stats[tran.DoiKhachId];

                    nha.SoTran++; khach.SoTran++;
                    nha.BanThang += tran.BanThangNha.Value;
                    nha.BanThua += tran.BanThangKhach.Value;
                    khach.BanThang += tran.BanThangKhach.Value;
                    khach.BanThua += tran.BanThangNha.Value;

                    if (tran.BanThangNha > tran.BanThangKhach)
                    {
                        nha.Thang++; nha.Diem += 3; khach.Thua++;
                    }
                    else if (tran.BanThangNha < tran.BanThangKhach)
                    {
                        khach.Thang++; khach.Diem += 3; nha.Thua++;
                    }
                    else
                    {
                        nha.Hoa++; nha.Diem++;
                        khach.Hoa++; khach.Diem++;
                    }
                }

                // FIFA Tie-breakers:
                // 1. Điểm → 2. Đối đầu trực tiếp → 3. Hiệu số → 4. Tổng bàn thắng
                var sorted = SapXepFIFA(stats.Values.ToList(), giai.TranDaus
                    .Where(t => t.BangId == bang.Id && t.TrangThai == "Closed").ToList());

                for (int i = 0; i < sorted.Count; i++)
                    sorted[i].ThuHang = i + 1;

                result[bang.Id] = sorted;
            }

            return result;
        }

        // ══════════════════════════════════════════════════════════
        // FIFA tie-breaker sắp xếp
        // ══════════════════════════════════════════════════════════
        private List<StandingRow> SapXepFIFA(
            List<StandingRow> rows, List<TranDau> tatCaTran)
        {
            // Group các đội cùng điểm
            var groups = rows.GroupBy(r => r.Diem)
                             .OrderByDescending(g => g.Key)
                             .ToList();

            var result = new List<StandingRow>();

            foreach (var group in groups)
            {
                var doiCungDiem = group.ToList();

                if (doiCungDiem.Count == 1)
                {
                    result.AddRange(doiCungDiem);
                    continue;
                }

                // Tính đối đầu trực tiếp trong nhóm
                var doiIds = doiCungDiem.Select(d => d.DoiId).ToHashSet();
                var tranDoiDau = tatCaTran.Where(t =>
                    doiIds.Contains(t.DoiNhaId) &&
                    doiIds.Contains(t.DoiKhachId)).ToList();

                var doiDauStats = doiCungDiem.ToDictionary(d => d.DoiId, _ => new {
                    diem = 0,
                    hieuSo = 0,
                    banThang = 0
                });

                // Tính điểm đối đầu trực tiếp
                var ddStats = doiCungDiem.ToDictionary(d => d.DoiId,
                    _ => new StandingSubRow());

                foreach (var t in tranDoiDau)
                {
                    if (!t.BanThangNha.HasValue) continue;
                    var nha = ddStats[t.DoiNhaId];
                    var khach = ddStats[t.DoiKhachId];

                    nha.BanThang += t.BanThangNha.Value;
                    nha.BanThua += t.BanThangKhach.Value;
                    khach.BanThang += t.BanThangKhach.Value;
                    khach.BanThua += t.BanThangNha.Value;

                    if (t.BanThangNha > t.BanThangKhach)
                    { nha.Diem += 3; }
                    else if (t.BanThangNha < t.BanThangKhach)
                    { khach.Diem += 3; }
                    else
                    { nha.Diem++; khach.Diem++; }
                }

                // Sắp xếp theo: Điểm ĐĐ → Hiệu số ĐĐ → BT ĐĐ → Hiệu số chung → BT chung
                var sorted = doiCungDiem.OrderByDescending(d => ddStats[d.DoiId].Diem)
                    .ThenByDescending(d => ddStats[d.DoiId].HieuSo)
                    .ThenByDescending(d => ddStats[d.DoiId].BanThang)
                    .ThenByDescending(d => d.HieuSo)
                    .ThenByDescending(d => d.BanThang)
                    .ToList();

                result.AddRange(sorted);
            }

            return result;
        }

        private class StandingSubRow
        {
            public int Diem { get; set; }
            public int BanThang { get; set; }
            public int BanThua { get; set; }
            public int HieuSo => BanThang - BanThua;
        }
    }

    // ── Public DTO dùng ở View ────────────────────────────────
    public class StandingRow
    {
        public int DoiId { get; set; }
        public string TenDoi { get; set; } = "";
        public string? LogoUrl { get; set; }
        public int ThuHang { get; set; }
        public int SoTran { get; set; }
        public int Thang { get; set; }
        public int Hoa { get; set; }
        public int Thua { get; set; }
        public int BanThang { get; set; }
        public int BanThua { get; set; }
        public int Diem { get; set; }
        public int HieuSo => BanThang - BanThua;
    }
}

File: tournament-service/src/main/java/com/pitchhub/tournament/service/StandingService.java
package com.pitchhub.tournament.service;

import com.pitchhub.tournament.dto.response.StandingRowResponse;
import com.pitchhub.tournament.model.BangDau;
import com.pitchhub.tournament.model.DoiBong;
import com.pitchhub.tournament.model.GiaiDau;
import com.pitchhub.tournament.model.TranDau;
import com.pitchhub.tournament.repository.GiaiDauRepository;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Tinh bang xep hang theo chuan FIFA tie-breakers.
 * Port truc tiep tu StandingService.cs — khong luu vao DB, tinh dong moi lan goi.
 */
@Service
public class StandingService {

    private final GiaiDauRepository giaiDauRepository;

    public StandingService(GiaiDauRepository giaiDauRepository) {
        this.giaiDauRepository = giaiDauRepository;
    }

    public Map<Integer, List<StandingRowResponse>> getStandings(Integer giaiDauId) {
        GiaiDau giai = giaiDauRepository.findDetailById(giaiDauId).orElse(null);
        if (giai == null) return new LinkedHashMap<>();

        Map<Integer, List<StandingRowResponse>> result = new LinkedHashMap<>();

        for (BangDau bang : giai.getBangDaus()) {
            List<DoiBong> doiTrongBang = giai.getDoiBongs().stream()
                    .filter(d -> bang.getId().equals(d.getBangId()))
                    .toList();

            Map<Integer, StandingRowResponse> stats = new LinkedHashMap<>();
            for (DoiBong d : doiTrongBang) {
                StandingRowResponse row = new StandingRowResponse();
                row.setDoiId(d.getId());
                row.setTenDoi(d.getTenDoi());
                row.setLogoUrl(d.getLogoUrl());
                stats.put(d.getId(), row);
            }

            List<TranDau> transClosed = giai.getTranDaus().stream()
                    .filter(t -> bang.getId().equals(t.getBangId()) && "Closed".equals(t.getTrangThai()))
                    .toList();

            for (TranDau tran : transClosed) {
                if (tran.getBanThangNha() == null || tran.getBanThangKhach() == null) continue;
                if (tran.getDoiNhaId() == null || tran.getDoiKhachId() == null) continue;
                if (!stats.containsKey(tran.getDoiNhaId()) || !stats.containsKey(tran.getDoiKhachId())) continue;

                StandingRowResponse nha = stats.get(tran.getDoiNhaId());
                StandingRowResponse khach = stats.get(tran.getDoiKhachId());

                nha.setSoTran(nha.getSoTran() + 1);
                khach.setSoTran(khach.getSoTran() + 1);
                nha.setBanThang(nha.getBanThang() + tran.getBanThangNha());
                nha.setBanThua(nha.getBanThua() + tran.getBanThangKhach());
                khach.setBanThang(khach.getBanThang() + tran.getBanThangKhach());
                khach.setBanThua(khach.getBanThua() + tran.getBanThangNha());

                if (tran.getBanThangNha() > tran.getBanThangKhach()) {
                    nha.setThang(nha.getThang() + 1);
                    nha.setDiem(nha.getDiem() + 3);
                    khach.setThua(khach.getThua() + 1);
                } else if (tran.getBanThangNha() < tran.getBanThangKhach()) {
                    khach.setThang(khach.getThang() + 1);
                    khach.setDiem(khach.getDiem() + 3);
                    nha.setThua(nha.getThua() + 1);
                } else {
                    nha.setHoa(nha.getHoa() + 1);
                    nha.setDiem(nha.getDiem() + 1);
                    khach.setHoa(khach.getHoa() + 1);
                    khach.setDiem(khach.getDiem() + 1);
                }
            }

            List<StandingRowResponse> sorted = sapXepFifa(new ArrayList<>(stats.values()), transClosed);
            for (int i = 0; i < sorted.size(); i++) {
                sorted.get(i).setThuHang(i + 1);
            }

            result.put(bang.getId(), sorted);
        }

        return result;
    }

    /** FIFA tie-breakers: Diem -> Doi dau truc tiep -> Hieu so -> Tong ban thang. */
    private List<StandingRowResponse> sapXepFifa(List<StandingRowResponse> rows, List<TranDau> tatCaTran) {
        Map<Integer, List<StandingRowResponse>> groups = rows.stream()
                .collect(Collectors.groupingBy(StandingRowResponse::getDiem, LinkedHashMap::new, Collectors.toList()));

        List<Integer> diemGiam = new ArrayList<>(groups.keySet());
        diemGiam.sort(Comparator.reverseOrder());

        List<StandingRowResponse> result = new ArrayList<>();

        for (Integer diem : diemGiam) {
            List<StandingRowResponse> doiCungDiem = groups.get(diem);

            if (doiCungDiem.size() == 1) {
                result.addAll(doiCungDiem);
                continue;
            }

            Set<Integer> doiIds = doiCungDiem.stream().map(StandingRowResponse::getDoiId).collect(Collectors.toSet());
            List<TranDau> tranDoiDau = tatCaTran.stream()
                    .filter(t -> t.getDoiNhaId() != null && t.getDoiKhachId() != null
                            && doiIds.contains(t.getDoiNhaId()) && doiIds.contains(t.getDoiKhachId()))
                    .toList();

            Map<Integer, int[]> ddStats = new HashMap<>(); // [diem, banThang, banThua]
            for (StandingRowResponse d : doiCungDiem) ddStats.put(d.getDoiId(), new int[]{0, 0, 0});

            for (TranDau t : tranDoiDau) {
                if (t.getBanThangNha() == null || t.getDoiNhaId() == null || t.getDoiKhachId() == null) continue;
                int[] nha = ddStats.get(t.getDoiNhaId());
                int[] khach = ddStats.get(t.getDoiKhachId());

                nha[1] += t.getBanThangNha();
                nha[2] += t.getBanThangKhach();
                khach[1] += t.getBanThangKhach();
                khach[2] += t.getBanThangNha();

                if (t.getBanThangNha() > t.getBanThangKhach()) {
                    nha[0] += 3;
                } else if (t.getBanThangNha() < t.getBanThangKhach()) {
                    khach[0] += 3;
                } else {
                    nha[0] += 1;
                    khach[0] += 1;
                }
            }

            List<StandingRowResponse> sorted = doiCungDiem.stream()
                    .sorted(Comparator
                            .comparingInt((StandingRowResponse d) -> ddStats.get(d.getDoiId())[0]).reversed()
                            .thenComparing(Comparator.comparingInt((StandingRowResponse d) ->
                                    ddStats.get(d.getDoiId())[1] - ddStats.get(d.getDoiId())[2]).reversed())
                            .thenComparing(Comparator.comparingInt((StandingRowResponse d) ->
                                    ddStats.get(d.getDoiId())[1]).reversed())
                            .thenComparing(Comparator.comparingInt(StandingRowResponse::getHieuSo).reversed())
                            .thenComparing(Comparator.comparingInt(StandingRowResponse::getBanThang).reversed()))
                    .toList();

            result.addAll(sorted);
        }

        return result;
    }
}
